guidelines/design-guidelines.md

```diff
@@ -31,4 +46,147 @@ Example:
+
+In order to achieve this goal a few key design decision have been made:
+
+### Folders to group capabilities
+
+Folders are used to group source files together to represent individual capabilities without affecting the namespaces of the components grouped together in these folders.
+
+For example, NServiceBus has a Recoverability capability folder, which is further divided into:
+
+* Faults
+* ImmediateRetries
+* DelayedRetries
+
+### Extension points on public APIs
+
+Public APIs provide extension points, allowing capabilities to hook in their business logic and float required state over those APIs to various extensions.
+
+For example, `SendOptions`, `PublishOptions` and `ReplyOptions` inherit from `ExtendableOptions` which provides a `ContextBag` to float additional state into the option classes. This state is then made available on the outgoing pipeline.
+
+### State belongs to a specific capability
+
+Instead of directly attaching state to composition roots and thereby exposing that state to other capabilities, we instead use the extension APIs to keep state internal to the capability.
+
+For example, let's consider `SendOptions` and `SendLocal`. Instead of exposing a `RouteToThisEndpoint` property on `SendOptions`
+
+```
+class SendOptions
+  public bool RouteToThisInstance { get; set; }
+```
+
+an extension method called `RouteToThisInstance` is used to set the internal state. The internal state belongs to the routing capability. Therefore the state cannot be attached to `SendOptions`. Since CSharp/.NET doesn't support extension properties the only way to implement this is to use an extension method. The benefit of extension methods is that the getter and the setter of an option of a capability can be implemented with different names, as opposed to properties.
+
+An example where we failed in the past to apply this is the `Headers` static class. It contains everything and the kitchen sink when it comes to headers.
+
+### Current categories
+
+This is the current list and will be expanded as needed
+
+* `Persistence` - Adapters for persistence infrastructure 
+* `Transport` - Adapters for queuing infrastructure
+* `Logging` - Adapters for logging frameworks
+* `Serialization` - Adapters for serializers
+* `Container` - Adapters for containers
+* `DataBus` - Adapters for data bus implementations
+* `Host` - Processes that host endpoints
+* `Encryption` - Adapters for encrypting messages/properties
+
+Note: This naming scheme may seen as redundant for "obvious" things like `NServiceBus.Logging.NLog` but not all intentions are obvious and we believe this redundancu is a price worth paying. This scheme also helps when searching the NuGet gallery (e.g. searching for containers using `NServiceBus.Container.*`).
```

Comment:
Comma after "`NServiceBus.Logging.NLog`".

Comma after "but not all intentions are obvious"